[6529] Remote configuration reload

epsilon-phase opened the pull request:
Adding support to trigging the configuration reload to
the control plane enables easier use in clusters or
multiple instance setups.

The current command doesn't support any options other than
the standard ones, and thus is invoked like
"irods-grid reload"

korydraughn commented:
You can use the testing environment to launch clusters of iRODS servers. That will allow you to verify the behavior of this code change.

`irods-grid` has options that allow you to target a specific set of hosts or all hosts. We need to verify that those options are honored when used.

korydraughn commented on server/main_server/src/main_server_control_plane.cpp:
```diff
@@ -567,9 +567,7 @@ namespace irods
-        // SIGHUP is the signal to reload the configuration. This shouldn't interact with it
-        signal(SIGHUP, SIG_IGN);
+        signal(SIGINT, ctrl_plane_handle_shutdown_signal);
         signal( SIGTERM, ctrl_plane_handle_shutdown_signal );
```
Remove the outer spaces around the arguments so that things line up with L570.

trel commented on scripts/irods/test/test_configuration_reload.py:
```diff
@@ -83,3 +84,27 @@ def test_how_long_restarting_takes(self):
                 self.admin.assert_icommand(
                     ['irule', rule, 'null', 'ruleExecOut'], "STDOUT", [str(i)])
+
+    def test_reload_command(self):
```
please name test as 'test_reload_command__issue_6529"

korydraughn commented on scripts/irods/test/test_configuration_reload.py:
```diff
@@ -83,3 +84,27 @@ def test_how_long_restarting_takes(self):
                 self.admin.assert_icommand(
                     ['irule', rule, 'null', 'ruleExecOut'], "STDOUT", [str(i)])
+
+    def test_reload_command(self):
+        server_config_filename = paths.server_config_path()
+        PROP_NAME = "is_issue_6529_done"
+        with lib.file_backed_up(server_config_filename):
+            with open(server_config_filename) as f:
+                svr_cfg = json.load(f)
+            svr_cfg[PROP_NAME] = "No?"
+            with open(server_config_filename, 'w') as f:
+                f.write(json.dumps(svr_cfg, sort_keys=True,
+                        indent=4, separators=(',', ': ')))
+            with session.make_session_for_existing_admin() as admin_session:
+                admin_session.environment_file_contents = IrodsConfig().client_environment
+                _, out, _ = admin_session.assert_icommand(
+                    "irods-grid reload --all", 'STDOUT_MULTILINE')
+                obj = json.loads(out)
+                changes = obj["hosts"][0]["configuration_changes_made"]
+                self.assertTrue(len(changes) == 1)
+                self.assertTrue(changes[0]["path"] == f"/{PROP_NAME}")
+                self.assertTrue(changes[0]["op"] == "add")
+                self.assertTrue(changes[0]["value"] == "No?")
+            rule = self.prepare_rule(PROP_NAME)
+            self.admin.assert_icommand(
+                ["irule", rule, 'null', 'ruleExecOut'], 'STDOUT', ["No?"])
```
Let's add some empty lines and comments so that it is easy to follow.

trel commented on scripts/irods/test/test_control_plane.py:
```diff
@@ -84,6 +84,27 @@ def test_shutdown_local_server(self):
             finally:
                 IrodsController().start()
 
+    def test_reload_command(self):
```
is this test... identical to the one in test_configuration_reload.py?

epsilon-phase replied:
Not exactly, it tests fewer things, so it should probably just live over there now

trel reviewed (commented):
consider renaming 'replace' with 'update'?   i don't see that code in this PR though...

korydraughn reviewed (changes requested):
Awesome catch on the call to `config_props_.patch`.

Try to think of ways in which this could break. We need to make sure we test the durability of the reload functionality.

korydraughn commented:
> consider renaming 'replace' with 'update'? i don't see that code in this PR though...

Where did you see `replace` appear?

trel commented:
In the manual demo - an entry that was changed in the config appeared in the output as ‘replace’ I think…

epsilon-phase commented:
add/replace/remove is defined by the json patch specification I think, so it'd be unwise to change that.

korydraughn commented:
Correct. Those words are coming from the nlohmann JSON library.

trel commented:
ooooh, i see.    carry on.

alanking commented on scripts/irods/test/test_configuration_reload.py:
```diff
@@ -83,3 +84,37 @@ def test_how_long_restarting_takes(self):
+
+            # Invoke irods-grid reload --all
+            with session.make_session_for_existing_admin() as admin_session:
+                admin_session.environment_file_contents = IrodsConfig().client_environment
```
I'm not sure that I understand the purpose of this line. Did we modify the client environment?

alanking commented on scripts/irods/test/test_configuration_reload.py:
```diff
@@ -83,3 +84,37 @@ def test_how_long_restarting_takes(self):
+            # Make the configuration changes
+            with open(server_config_filename) as f:
+                svr_cfg = json.load(f)
+            svr_cfg[PROP_NAME] = "No?"
```
"No?" -> "Yes?" :p

alanking commented on scripts/irods/test/test_configuration_reload.py:
```diff
@@ -83,3 +84,37 @@ def test_how_long_restarting_takes(self):
+                self.assertTrue(changes[0]["op"] == "add")
+                self.assertTrue(changes[0]["value"] == "No?")
+
+            # Also make sure that the actual
```
The suspense is killing me...

alanking reviewed (commented):
Looks good! Just one real comment

alanking reviewed (approved):
Looks good to me. @trel?

trel commented:
clang-tidy has a couple opinions that look real - should those be addressed here?

otherwise, please # it.

epsilon-phase commented:
I addressed those concerns as best as I can, the missing file it complains seems like it probably isn't correct about, since that wouldn't compile otherwise

trel commented:
I think we did it!

alanking commented:
I agree. clang-tidy is a work in progress from a configuration perspective so I think the remaining issues are invalid. Let's # it.

alanking commented:
I see that this has already been done. Merging
